README.md
# steamworks4j

A thin wrapper which allows Java applications to access the Steamworks C++ API.

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Github: Latest Version](https://img.shields.io/github/v/release/Zweistein2/steamworks4j?sort=semver)](https://github.com/Zweistein2/steamworks4j/releases)
[![Github: Releases](https://img.shields.io/github/downloads/Zweistein2/steamworks4j/total)](https://github.com/Zweistein2/steamworks4j/releases)
[![Sonar: Alert Status](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=alert_status)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)

---

Sonar Analysis (master):

[![Sonar: Coverage](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=coverage)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Bugs](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=bugs)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Vulnerabilities](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=vulnerabilities)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Code Smells](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=code_smells)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Lines of Code](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=ncloc)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Technical Debt](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=sqale_index)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Maintainability Rating](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=sqale_rating)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Reliability Rating](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=reliability_rating)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)
[![Sonar: Security Rating](https://sonar.potionlabs.de/api/project_badges/measure?project=steamworks4j:master&metric=security_rating)](https://sonar.potionlabs.de/dashboard?id=steamworks4j%3Amaster)

### Documentation
[http://code-disaster.github.io/steamworks4j/](http://code-disaster.github.io/steamworks4j/)

### Games using steamworks4j
[http://code-disaster.github.io/steamworks4j/showcase.html](http://code-disaster.github.io/steamworks4j/showcase.html)

### A note about feature requests and contributions

The wrapper does not implement all Steamworks interfaces and functions. If you miss something, feel free to [contact me](mailto:fkarolat@potionlabs.de), or just open a [new issue](https://github.com/Zweistein2/steamworks4j/issues) right here on GitHub. Of course, pull requests are welcome too.


CHANGES.md
## change log

### HEAD
- Updated to Steamworks SDK v1.49.
- Updated to Java 1.14.
- Complete overhaul of the internal structure.
- Added SteamInventory interface. (#92)
- Added `SteamUserCallback.onAuthSessionTicket()`. (#83)

### [1.8.0]
- MacOS: removed 32-bit support.
- MacOS: raised minimum version support to OS X 10.9.
- API change: split up initialization into `SteamAPI.loadLibraries()` and `SteamAPI.init()`. Allows to call some native functions, like restartAppIfNecessary(), before SteamAPI.init().
- API change: applied same change as above to `SteamGameServerAPI.loadLibraries()` and `SteamGameServerAPI.init()`.
- SteamGameServerAPI skips loading of shared libraries if already loaded by SteamAPI. (#56)
- Added `SteamUGC.deleteItem()`. (#64)
- Added Voice API. [contributed by sf17k]
- Fixed signature of `SteamMatchmaking.getLobbyMemberLimit()`. (#67)
- Added `SteamFriends.getFriendGamePlayed()`. (#68)
- Added `SteamMatchmaking.[set|get]LobbyMemberData()`. (#72) [contributed by XaeroDegreaz]
- Added `SteamUserStats.downloadLeaderboardEntriesForUsers()` (#74)

### [1.7.0]
- Updated to Steamworks SDK v1.42.
- Reviewed all API functions which use NIO buffers. They now all use direct *byte* buffers and properly read `ByteBuffer#position()` as offset and `ByteBuffer#remaining()` as size/capacity markers. Also, they never change the buffer state, so for functions returning data, the caller must ensure that buffer position and/or limit are updated accordingly.
- Removed redistributable shared libraries from source control. They are now pulled from the `./sdk/redistributable_bin/` folder directly by Maven.
- Added `steamworks4j-server` module. It creates two native libraries, steamworks4j-server and steamworks4j-encryptedappticket.
- Added `tests` module, which now contains all test apps.
- Added SteamRemoteStorage callbacks to notify application about external changes to workshop item subscriptions. (#42)
- Added SteamRemoteStorage functions for asynchronous file access and file/cloud queries.
- Added SteamNetworking.getP2PSessionState(). (#43)
- Added SteamUser functions to request encrypted app ticket. (#45)
- Added SteamEncryptedAppTicket server functions. (#45)
- API change: adjusted SteamGameServer.sendUserConnectAndAuthenticate() to better mimic the C++ API.
- Added SteamMatchmakingServers wrapper.
- Refactored shared library loader to work with the new `server` and `test` modules.
- Added some options to customize the shared library loader behavior during development.
- Moved all SteamGameServer* classes to the server module.
- API change: Removed `API` parameter when creating HTTP and networking wrappers. Added SteamGameServerHTTP and SteamGameServerNetworking to create server-side versions.
- Added SteamFriendsCallback.onGameServerChangeRequested(). (#48)
- Removed support for Linux 32-bit.
- Updated native Windows build to a newer version of premake 5 (alpha 11).
- Added some missing SteamApps functions. (#50)
- Added SteamScreenshots wrapper.
- All Maven modules are now compiled with Java 7.
- Added SteamUtils.getImageSize(). (#62)
- Added com.codedisaster.steamworks.Version to query library version at runtime.

### [1.6.2]
- Cleaned up Maven project structure. Removed superfluous releases of 'steamworks4j-parent' and 'steamworks4j-jnigen' repositories.
- Reduced JNI overhead for invocations of callback adapters.
- Fixed crash when trying to run multiple processes using steamworks4j simultaneously.

### [1.6.1]
- Updated to Steamworks SDK v1.39.
- Added SteamFriends.setPersonaName().
- Added remaining functions to activate Steam overlay.
- Added functions for rich presence and game invites to SteamFriends (#38).
- API change: the SteamFriendsCallback interface has been extended to reflect the latest additions.
- API change: removed SteamID.getNativeHandle(). All native handles can now be queried with SteamNativeHandle.getNativeHandle().
- The default behaviour of SteamAPI.isSteamRunning() has changed to *not* call SteamAPI_IsSteamRunning(), see JavaDocs for more information.
- Added SteamAPI.releaseCurrentThreadMemory().

### [1.6.0]
- Added SteamController interface.
- Fixed resource files (prebuilt native libraries) included in source jars.

### [1.5.0]
- Updated to Steamworks SDK v1.38a.
- Added SteamUserCallback.onMicroTxnAuthorization() callback.
- API change: renamed SteamUserCallback.onValidateAuthTicketResponse() to SteamUserCallback.onValidateAuthTicket().
- API change: all uses of appID in member functions and callbacks now use 32-bit ints.

### [1.4.0]
- Removed steamworks4j-natives subproject. Native libraries are now added as resources to steamworks4j.jar directly.
- Added functions to query global stats. This also adds a new method to SteamUserStatsCallback (#36).
- Updated shared library loader, adding a few fallbacks in case executables cannot be written to, or launched from, java.io.tmpdir.
- API change: SteamAPI.init() now throws SteamException if loading of native libraries fails. The function now returns false only if initialization fails for Steamworks reasons (e.g. no Steam client running).

### [1.3.1]
- Updated to Steamworks SDK v1.37.
- Added some missing SteamUtils functions (#35).
- Added more fields returned in SteamUGCDetails (#34).

### [1.3.0]
- Updated to Steamworks SDK v1.36.
- API change: changed auth ticket API to use SteamAuthTicket objects instead of plain integer handles.
- API change: getAuthSessionTicket() now takes an int[1] argument to return the the required buffer size.
- Added buffer validation checks to auth ticket functions.
- Added auth ticket test functions to SteamNetworkingTestApp.
- Added two missing SteamRemoteStorage.updatePublishedFile*() functions.
- Added public functions to serialize SteamID as plain 64 bit value.
- Added ISteamUGC-based workshop functions.
- API change: SteamUtils now exposes a SteamUtilsCallback interface.
- Added SteamUtils.onSteamShutdown() callback.

java-wrapper/src/main/java/com/codedisaster/steamworks/SteamResult.java
package com.codedisaster.steamworks;

public enum SteamResult {

	/**
	 * Error codes. Values need to map directly to SDK error codes!
	 */
	OK(1),
	Fail(2),
	NoConnection(3),
	//NoConnectionRetry(4),
	InvalidPassword(5),
	LoggedInElsewhere(6),
	InvalidProtocolVer(7),
	InvalidParam(8),
	FileNotFound(9),
	Busy(10),
	InvalidState(11),
	InvalidName(12),
	InvalidEmail(13),
	DuplicateName(14),
	AccessDenied(15),
	Timeout(16),
	Banned(17),
	AccountNotFound(18),
	InvalidSteamID(19),
	ServiceUnavailable(20),
	NotLoggedOn(21),
	Pending(22),
	EncryptionFailure(23),
	InsufficientPrivilege(24),
	LimitExceeded(25),
	Revoked(26),
	Expired(27),
	AlreadyRedeemed(28),
	DuplicateRequest(29),
	AlreadyOwned(30),
	IPNotFound(31),
	PersistFailed(32),
	LockingFailed(33),
	LogonSessionReplaced(34),
	ConnectFailed(35),
	HandshakeFailed(36),
	IOFailure(37),
	RemoteDisconnect(38),
	ShoppingCartNotFound(39),
	Blocked(40),
	Ignored(41),
	NoMatch(42),
	AccountDisabled(43),
	ServiceReadOnly(44),
	AccountNotFeatured(45),
	AdministratorOK(46),
	ContentVersion(47),
	TryAnotherCM(48),
	PasswordRequiredToKickSession(49),
	AlreadyLoggedInElsewhere(50),
	Suspended(51),
	Cancelled(52),
	DataCorruption(53),
	DiskFull(54),
	RemoteCallFailed(55),
	PasswordUnset(56),
	ExternalAccountUnlinked(57),
	PSNTicketInvalid(58),
	ExternalAccountAlreadyLinked(59),
	RemoteFileConflict(60),
	IllegalPassword(61),
	SameAsPreviousValue(62),
	AccountLogonDenied(63),
	CannotUseOldPassword(64),
	InvalidLoginAuthCode(65),
	AccountLogonDeniedNoMail(66),
	HardwareNotCapableOfIPT(67),
	IPTInitError(68),
	ParentalControlRestricted(69),
	FacebookQueryError(70),
	ExpiredLoginAuthCode(71),
	IPLoginRestrictionFailed(72),
	AccountLockedDown(73),
	AccountLogonDeniedVerifiedEmailRequired(74),
	NoMatchingURL(75),
	BadResponse(76),
	RequirePasswordReEntry(77),
	ValueOutOfRange(78),
	UnexpectedError(79),
	Disabled(80),
	InvalidCEGSubmission(81),
	RestrictedDevice(82),
	RegionLocked(83),
	RateLimitExceeded(84),
	AccountLoginDeniedNeedTwoFactor(85),
	ItemDeleted(86),
	AccountLoginDeniedThrottle(87),
	TwoFactorCodeMismatch(88),
	TwoFactorActivationCodeMismatch(89),
	AccountAssociatedToMultiplePartners(90),
	NotModified(91),
	NoMobileDevice(92),
	TimeNotSynced(93),
	SmsCodeFailed(94),
	AccountLimitExceeded(95),
	AccountActivityLimitExceeded(96),
	PhoneActivityLimitExceeded(97),
	RefundToWallet(98),
	EmailSendFailure(99),
	NotSettled(100),
	NeedCaptcha(101),
	GSLTDenied(102),
	GSOwnerDenied(103),
	InvalidItemType(104),
	IPBanned(105),
	GLSTExpired(106),
	InsufficientFunds(107),
	TooManyPending(108),
	NoSiteLicensesFound(109),
	WGNetworkSendExceeded(110),
	AccountNotFriends(111),
	LimitedUserAccount(112),
	CantRemoveItem(113),
	AccountDeleted(114),
	ExistingUserCancelledLicense(115),
	CommunityCooldown(116),

	/**
	 * If this is returned(), we missed to "port" an Steam error code above.
	 */
	UnknownErrorCode_NotImplementedByAPI(0);

	private final int code;
	private static final SteamResult[] valuesLookupTable;

	SteamResult(final int code) {
		this.code = code;
	}

	public static SteamResult byValue(final int resultCode) {
		if (resultCode < valuesLookupTable.length) {
			return valuesLookupTable[resultCode];
		} else {
			return UnknownErrorCode_NotImplementedByAPI;
		}
	}

	static {
		final SteamResult[] values = values();
		int maxResultCode = 0;

		for (final SteamResult value : values) {
			maxResultCode = Math.max(maxResultCode, value.code);
		}

		valuesLookupTable = new SteamResult[maxResultCode + 1];

		for (final SteamResult value : values) {
			valuesLookupTable[value.code] = value;
		}
	}
}
